prqlc/prql-compiler/src/sql/gen_expr.rs

```diff
@@ -104,6 +104,32 @@ pub(super) fn translate_expr(expr: Expr, ctx: &mut Context) -> Result<ExprOrSour
                 }
                 "std.concat" => return Ok(process_concat(&expr, ctx)?.into()),
                 "std.array_in" => return Ok(process_array_in(args, ctx)?.into()),
+                "std.date.to_string" => {
+                    if let [date_format_exp @ Expr {
+                        kind: ExprKind::Literal(Literal::String(date_format)),
+                        ..
+                    }, col_expr] = &args[..]
```

Comment:
This unpacking is a crucial detail. It assumes that the format is a string literal.

It might have been defined in a separate variable, but it will not work if you do:
```
d | date.to_string ('%Y' + '%M')
```
(I'm using `+` as string concatenation here).

I think this restriction is ok for now (since we do have the same restriction on `from_text` and a few others), but I'd be nice if we'd have a plan for the future:

If this is some expression that we cannot statically evaluate in the compiler, we need to parse it in the database. With SQL. On different databases. This will be hard to do.

Probably the best approach to do this would be to write this parsing and mapping in PRQL.

Reply:
Yes good point.

Though how would we do this in PRQL directly?

@aljazerzen do you think it's feasible to have the function take the "materialized" value of an argument while being in rust? So `d | date.to_string (f"%{foo}-%M")`, where `foo="%Y"` would work?

Reply:
I didn't understand how you plan to do that sorry

Reply:
I also don't understand what you mean @max-sixty 

Let me try to explain the problem on another example. Say that we have this query:
```
from dates_to_display
select {my_date, my_format}
select {std.date.to_string my_date my_format}
```
The problem is now `my_format`, which is a string column in the database that we have no insight into at compile time.

If we want to allow usages like this, we must implement the conversion from `Year: %Y` to `TO_CHAR('Year: YYYY', ...)` in SQL. So we must parse the fstrtime format and produce the database's format string using what's available in the target database's dialect.

Because this 'parsing and producing' code is complex, we don't want to write or maintain it for each database separately. Instead, we could implement it using PRQL. The problem here is that PRQL is not powerful enough to express such algorithms with ease.

---

In any case, looking at strftime format, it seems like it is the most widespread time-formatting language, so I don't object adding it into PRQL's std module, regardless of the cost of the full implementation.

Reply:
Sorry if I'm being confusing.

I'm also a bit confused!

> The problem is now `my_format`, which is a string column in the database that we have no insight into at compile time.

The date format string needs to be in the query itself, in the SQL dialects I'm familiar with. We can't do:

```sql
SELECT 
  dt, 
  TO_CHAR(dt, fmt_column) as dt_str   # fmt_column is a column
FROM artists
```

---

The way that we _can_ get a dynamic date format string is by forming it in PRQL:

```elm
let dt_string = "%Y-%m" 

from artists
select {dt, dt_str = (dt | date.to_string dt_string)}
```

I (I think mistakenly) thought you were claiming that wouldn't work, and was asking if it were possible. After reading the messages, I think we both believe this is possible.

---

Maybe if there's still confusion, would you mind clarifying the example that wouldn't work?